### ISSUE TITLE
Generate new method inside existing impl block if possible

### ISSUE THREAD

**Issue by ortem** [intellij-rust/intellij-rust#6660]
## Environment

* **IntelliJ Rust plugin version:** 0.3.140.SNAPSHOT-203-dev
* **Rust toolchain version:** 1.48.0 (7eac88abb 2020-11-16) x86_64-apple-darwin
* **IDE name and version:** CLion 2020.3.1 (CL-203.6682.181)
* **Operating system:** macOS 10.16
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
`CreateFunctionIntention` should generate a new method inside an existing impl block if possible. Actually, there is such logic in `org.rust.ide.intentions.CreateFunctionIntention#getOrCreateImpl`, but probably it contains some bugs and does not work properly in case of public method calls.

<img width="348" alt="Screenshot 2021-01-13 at 18 32 11" src="https://user-images.githubusercontent.com/4854600/104474539-29a1e880-55cf-11eb-9957-b115ab1a6d1c.png">
<img width="367" alt="Screenshot 2021-01-13 at 18 32 16" src="https://user-images.githubusercontent.com/4854600/104474545-2ad31580-55cf-11eb-80e1-0a558b5125a7.png">

## Steps to reproduce

Before:
```Rust
struct S;

impl S {
    fn foo(&self) {}
}

fn main() {
    let s = S;
    s.foo();
    s.bar/*caret*/(); // Invoke "Create method" intention
}
```

After:
```Rust
struct S;

impl S {
    pub(crate) fn bar(&self) {
        unimplemented!()
    }
}

impl S {
    fn foo(&self) {}
    // `bar` implementation could be here
}

fn main() {
    let s = S;
    s.foo();
    s.bar();
}
```

**Comment by Kobzol** [comment]
This is currently intended. `getOrCreateImpl` only works for cases where the method call itself is inside an `impl`. If it isn't we would have to invoke a (potentially global) search for impls from an intention, which could have been expensive (I think that I remember that @vlad20012 warned me about this in some earlier intention that also created new impl items). @vlad20012 wdyt?

**Comment by vlad20012** [comment]
> In theory, this can be done by `ImplLookup.findImplsAndTraits(ty).filterIsInstance<TraitImplSource.ExplicitImpl>().filter { it.isInherent }`. But there can be multiple suitable impls

(from https://github.com/intellij-rust/intellij-rust/pull/5505#issuecomment-682208398)

I's say it's possible, but here can be multiple suitable impls, including blanket and impls with different trait bounds (`impl<T: Foo> Bar<T> {}`), and anyway there can be multiple simple inherent impls (`impl S {} impl S {}`). If we handle these cases correctly and decide what to do with multiple absolutely suitable impls, this issue can be solved

**Comment by vlad20012** [comment]
I can’t remember that I warned someone about impl search expensiveness in intentions. Maybe that was a more special case...

**Comment by Kobzol** [comment]
Right, I got it mixed up a little bit, sorry. I was indeed referring to the comment that you have posted, along with discussion in https://github.com/intellij-rust/intellij-rust/pull/6037#discussion_r480306331.

The problem is that there can indeed be multiple `impl`s, with different generic parameters, bounds, etc. For example:
```rust
struct S<T>(T);
trait Trait {}

impl<T> S<T> {}
impl<R: Trait> S<R> {}
impl S<u32> {}
```

Probably the best thing would be to insert the method in the "most general" impl, or in an impl that most matches the current type parameters of the method receiver (e.g. if I called `s.foo()` on `S<u32>`, it could create the method on the last impl). I don't really know how to do that generally.

Maybe a simpler heuristic could solve most issues, i.e. if there is only a single existing `impl`, use it, if not, create a new one?

**Comment by nschoellhorn** [comment]
Hey everyone, I was looking into the code today and wondered why it was done the way it currently is. This issue explains a lot, thank you for that!
Since this issue is pretty annoying (at least for me :D), I feel like the suggestion in your last comment, @Kobzol, is pretty good for now. Of course, having the IDE place it into the correct specialization would be pretty neat but for now, placing it into the existing `impl` if there's only one would be enough as far as I can tell. I'd be more than happy to submit a PR for that if we can agree on that solution.

**Comment by Kobzol** [comment]
@vlad20012 what do you think about inserting the function inside an existing `impl` if there is only one `impl`?
